svj_ntuple_processing/__init__.py

```diff
@@ -763,151 +745,117 @@ def filter_preselection(array, single_muon_cr=False):
-     met, metphi
-     )
+    mt = calculate_mt(pt, eta, phi, e, met, metphi)
     # compute but don't apply
-    cutflow['180<mt<650'] = np.sum((mt>180) & (mt<650))
-
-    cutflow['preselection'] = len(a)
-
-    copy.array = a
-    logger.debug('cutflow:\n%s', pprint.pformat(copy.cutflow))
-    return copy
-
-def selection_plots(array):
-    # to make the n-1 plots
-    copy = array.copy()
-    a = copy.array
-    cutflow = copy.cutflow
-
-    # At least 2 AK15 jets
-    a = a[ak.count(a['JetsAK15.fCoordinates.fPt'], axis=-1) >= 2]
-    cutflow['n_ak15jets>=2'] = len(a)
-    # jetid for AK15 jets
-    a = a[a['JetsAK15_ID'][:,1]>0]
-    cutflow['ak15jets_id'] = len(a)
-    # subleading eta < 2.4 eta
-    a = a[np.abs(a['JetsAK15.fCoordinates.fEta'][:,1])<2.4]
-    cutflow['subl_eta<2.4'] = len(a)
-
-    # MET filters
-    for b in met_filters:
-        a = a[a[b]!=0] # Pass events if not 0, is that correct?
-    cutflow['metfilter'] = len(a)
+    arrays.cutflow['180<mt<650'] = np.sum((mt>180) & (mt<650))
+    return arrays
 
-    # At least 2 AK4 jets --> deadcells study
-    a = a[ak.count(a['Jets.fCoordinates.fPt'], axis=-1) >= 2]
-    cutflow['n_ak4jets>=2'] = len(a)
 
-    # Filter out jets that are too close to dead cells
-    ak4jet_eta = a['Jets.fCoordinates.fEta'][:,1].to_numpy()
-    ak4jet_phi = a['Jets.fCoordinates.fPhi'][:,1].to_numpy()
-    dead_cell_mask = veto_phi_spike(
-        dataqcd_eta_ecaldead[array.year], dataqcd_phi_ecaldead[array.year],
-        ak4jet_eta, ak4jet_phi,
-        rad = 0.01
-        )
+class EventSelector():
+    def __init__(self, name:str, filter_list: List[Callable]):
+        self.name = name
+        self.filter_list = filter_list
 
-    a = a[dead_cell_mask]
-    cutflow['ecaldeadcells'] = len(a)
+    def __call__(self, arrays:Arrays):
+        copy = arrays.copy() # Operate on the copy
+        for filter in filter_list:
+            copy = filter(copy)
+        copy.cutflow[self.name] = len(copy.array)
+        logger.debug('cutflow:\n%s', pprint.pformat(copy.cutflow))
+        return copy
 
-    cutflow['selection_cuts'] = len(a)
 
-    copy.array = a
-    logger.debug('cutflow:\n%s', pprint.pformat(copy.cutflow))
-    return copy
+filter_preselection = EventSelector("preselection", [
+    select_jet_topology,
+    select_trigger,
+    select_trigger_object,
+    select_positive_ecf,
+    select_rt,
+    select_muon_veto,
+    select_electron_veto,
+    select_metdphi,
+    select_metfilter_standard,
+    select_highpt_muon_veto,
+    select_metfilter_custom,
+    select_mt,
+])
+
+def filter_preselection_minus_one(cut_type):
+    # The selection syntax would be something like:
+    # filter_preselection_minus_one("metdphi")(array)
+    assert any([m.endswith(cut_type) for m in filter_preselection.filter_list])
+    return EventSelector(
+            name=f"preselection_minus_{cut_type}",
+            [m for m in filter_preselection.filter_list if m.__name__.endswith(cut_type)]
```

Comment:
It would be safer to do:
```python
[m for m in filter_preselection.filter_list if m.__name__==f"select_{cut_type}"]
```
otherwise e.g. `filter_preselection_minus_one("veto")` would knock out 3 cuts...